### ISSUE TITLE
Unreleased Resource: Database

### ISSUE THREAD

**Issue by QiAnXinCodeSafe** [OpenRefine/OpenRefine#4080]
https://github.com/OpenRefine/OpenRefine/blob/409cf2c30097ae7a5981e941d4bbee49896cd590/extensions/database/src/com/google/refine/extension/database/mariadb/MariaDBDatabaseService.java#L85-L86

The program can potentially fail to release a database resource.

**Comment by vdimopoulos** [comment]
Hi there,

A friend and I are looking to contribute on an Open Source project for our Undergraduate Studies. 
We have read the documentation and debugged OpenRefine locally a few times.

Could we please handle this issue?

Thanks in advance c:

**Comment by thadguidry** [comment]
@vdimopoulos Yes, the general practice for releasing database objects in Java is to use try-with-resources.  The current architecture is Import only, I.E. SQL Query only.  So we only need connection pools in order to perform a query against them to import a table containing data.  Query failure handling should already be wired up well enough, I think, but you can double check that as well I would say.

**Comment by vdimopoulos** [comment]
Thank you for your feedback.

Indeed Connection Pooling is used in the executeQuery() method, which calls the shutdown() method on the MariaDBConnectionManager  to call close() on the connection. Would that mean that this issue is actually solved, since the Connection resource is actually closed in the finally block?

We also notice that Connection Pooling is used in the getRows() method of the class, but there the connection is not getting closed anywhere. Do you think that this is something that should be handled or happens on purpose in order to leave the connection open for executeQuery() to be called later on?

Other than that, getMetadata() and getColumns() seem to create a new connection and import in the Pool, so there should be no issue.

Once again, we appreciate your time.

**Comment by vdimopoulos** [comment]
After giving the issue another glance, could it only be that we need to close the Statement and ResultSet resources that are staying open and leaving the Connection as is?

**Comment by thadguidry** [comment]
@vdimopoulos Likely, yes, and you should be able to test if they are staying open after a successful query is performed (even if manually by monitoring MariaDB itself).
ResultSet brings about an interesting question there... is there ever the potential for queryResult to be NULL and if so, what would be the SQLException? :-)  Something else to look at and think about perhaps in another issue.  But no matter, if any exception when attempting a connection or query, we seem to `finally` always shutdown the connection instance, so that makes sense.

**Comment by glagou** [comment]
Hey @thadguidry, I am the Vasilis colleague. We understood the issue and we will be working on this very soon. Thanks a lot for the guidance c:

**Comment by glagou** [comment]
Hey @thadguidry, 

We set up a local MariaDB instance and looked into the issue locally. We changed the code to include the Statement and ResultSet objects into a try-with-resources block, as shown into the screenshots below and everything seems to be working fine.

There are three things that we would like to discuss:
1) We have made the above changes into all the methods of the class, since the issue was not present only in the executeQuery() method that this #4080 describes.
2) The same thing seems to happen to the rest DatabaseService classes where the code is quite similar. If this change is correct, maybe we should apply it to the rest of the classes before making a Pull Request?
3) We could not find a way to call the executeQuery() method. Is it not used anymore? Could you please tell us if and how this method is called?

Thanks in advance. :smile:

**Comment by thadguidry** [comment]
@glagou For 3. In the OpenRefine application you should be able to Create Project and use the SQL importer which allows you to make the DB connection and then query a chosen table to import the data into the preview and then create the project.

For 1. and 2. I'm not sure and don't have time to dig in to refresh my memory, sorry.

**Comment by glagou** [comment]
We have tested both a MySQL and a MariaDB instance and the executeQuery() method seems to never get called. By SQL importer you mean the Database tab that is used to import records by executing SQL queries, right? We have attached a screenshot displaying where we execute the queries.

![image](https://user-images.githubusercontent.com/72858319/163856520-838fca58-fe42-4915-b975-122892024d15.png)

**Comment by glagou** [comment]
As for (1) and (2), we could make a Pull Request for (1) and probably create a new issue regarding (2). Whatsoever, if the executeQuery() method should get called when following the steps above, that does not seem to be happening right now.

**Comment by thadguidry** [comment]
@glagou Then I'm not sure about the executeQuery() without poking around. Sorry, busy. But I'm sure you can figure it out or maybe it was something leftover when Tony created the extension specifically for testing or something.
